tests/integrations/server/kv_service.rs

```diff
@@ -212,3 +195,49 @@ fn test_read_index_check_memory_locks() {
     assert!(!resp.has_locked());
     assert_eq!(cm.max_ts(), start_ts);
 }
+
+#[test]
+fn test_prewrite_check_max_commit_ts() {
+    let mut cluster = new_server_cluster(0, 1);
+    cluster.run();
+
+    let cm = cluster.sim.read().unwrap().get_concurrency_manager(1);
+    cm.update_max_ts(100.into());
+
+    let (client, ctx) = build_client(&cluster);
+
+    let mut req = PrewriteRequest::default();
+    req.set_context(ctx.clone());
+    req.set_primary_lock(b"k1".to_vec());
+    let mut mutation = Mutation::default();
+    mutation.set_op(Op::Put);
+    mutation.set_key(b"k1".to_vec());
+    mutation.set_value(b"v1".to_vec());
+    req.mut_mutations().push(mutation);
+    req.set_start_version(10);
+    req.set_max_commit_ts(200);
+    req.set_lock_ttl(20000);
+    req.set_use_async_commit(true);
+    let resp = client.kv_prewrite(&req).unwrap();
+    assert_eq!(resp.get_min_commit_ts(), 101);
+
+    let mut req = PrewriteRequest::default();
+    req.set_context(ctx.clone());
+    req.set_primary_lock(b"k2".to_vec());
+    let mut mutation = Mutation::default();
+    mutation.set_op(Op::Put);
+    mutation.set_key(b"k2".to_vec());
+    mutation.set_value(b"v2".to_vec());
+    req.mut_mutations().push(mutation);
+    req.set_start_version(20);
+    req.set_max_commit_ts(50);
+    req.set_lock_ttl(20000);
+    req.set_use_async_commit(true);
+    let resp = client.kv_prewrite(&req).unwrap();
+    assert_eq!(
+        resp.get_errors()[0]
+            .get_commit_ts_too_large()
+            .get_commit_ts(),
+        101
+    );
```

Comment:
Can you also test that the memory lock is not locked after returning the error?

Reply:
OK. I think it's more meaningful to check this when some key is locked successfully. So this is checked by another failpoint test. You can take a look.